// @aws-cdk/aws-stepfunctions-tasks/lib/apigateway/base-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthType = exports.HttpMethod = void 0;
/** Http Methods that API Gateway supports */
var HttpMethod;
(function (HttpMethod) {
    /** Retreive data from a server at the specified resource */
    HttpMethod["GET"] = "GET";
    /** Send data to the API endpoint to create or udpate a resource */
    HttpMethod["POST"] = "POST";
    /** Send data to the API endpoint to update or create a resource */
    HttpMethod["PUT"] = "PUT";
    /** Delete the resource at the specified endpoint */
    HttpMethod["DELETE"] = "DELETE";
    /** Apply partial modifications to the resource */
    HttpMethod["PATCH"] = "PATCH";
    /** Retreive data from a server at the specified resource without the response body */
    HttpMethod["HEAD"] = "HEAD";
    /** Return data describing what other methods and operations the server supports */
    HttpMethod["OPTIONS"] = "OPTIONS";
})(HttpMethod = exports.HttpMethod || (exports.HttpMethod = {}));
/**
 * The authentication method used to call the endpoint
 */
var AuthType;
(function (AuthType) {
    /** Call the API direclty with no authorization method */
    AuthType["NO_AUTH"] = "NO_AUTH";
    /** Use the IAM role associated with the current state machine for authorization */
    AuthType["IAM_ROLE"] = "IAM_ROLE";
    /** Use the resource policy of the API for authorization */
    AuthType["RESOURCE_POLICY"] = "RESOURCE_POLICY";
})(AuthType = exports.AuthType || (exports.AuthType = {}));
